src/dataFiles.ts

```diff
@@ -39,65 +36,104 @@ export async function initializeDataDirs(): Promise<void> {
 	});
 	if (!dataDirs?.length) return;
 	for (const dataDir of dataDirs) {
-		watchers.set(dataDir, null);
+		modifiedPaths.set(dataDir, new Set());
+		watchers.set(
+			dataDir,
+			watch(dataDir, { recursive: true, persistent: false }, (_, f) => {
+				if (!f) return;
+				const fullPath = resolve(join(dataDir, f));
+				if (fullPath === resolve(dataDir)) return;
+				modifiedPaths.get(dataDir)!.add(fullPath);
+			}),
+		);
 	}
 	logger.info("Indexing dataDirs for reverse lookup...");
-	await indexDataDirs();
+	await indexDataPaths(findSearcheesFromAllDataDirs());
+}
+
+/**
+ * This gets called on each rss/webhook/announce event.
+ */
+export async function indexDataDirs(): Promise<void> {
+	const { maxDataDepth } = getRuntimeConfig();
+	await Promise.all(
+		Array.from(watchers.keys()).map((dataDir) => {
+			const modified = modifiedPaths.get(dataDir)!;
+			const eventPaths: string[] = [];
+			while (modified.size) {
+				const path: string | undefined = modified.values().next().value;
+				if (!path) continue;
+				if (!modified.delete(path)) continue;
+				eventPaths.push(path);
+			}
```

Comment:
is this equivalent to `const eventPaths = Array.from(modified).filter(isTruthy)`?

Reply:
We need to remove the item from the set as we will be reindexing it. I'm using the iterator as JS does not have `.pop()` on sets. The goal here is to go through all the modified paths in a concurrent safe way. As multiple indexes could be triggered at the same time from announces or rss.